### PR TITLE
Inline edit/remove of links in HTML editor

### PR DESCRIPTION
**Pull request by phalkunz**
Issue: https://github.com/silverstripe/silverstripe-admin/issues/66

### CODE REVIEW

File: client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    return this;
+  }
+}
+
+/**
+ * @param  tinymce selection object
+ * @param  {Boolean}
+ * @param  {Array} array of tag names - e.g. "A"
+ */
+function shouldShowToolbar(selection, isEditorFocused, tagTypes) {
```
Comment (flamerohr) [inline client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js]: include this in the exports as well

File: client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    this.setStyles({ left, top });
+
+    return this;
+  }
+}
+
+/**
+ * @param  tinymce selection object
+ * @param  {Boolean}
+ * @param  {Array} array of tag names - e.g. "A"
```
Comment (flamerohr) [inline client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js]: almost right with the JSDocs :D
` * @param {Object} selection tinymce selection object`
where the first selection is the name of the param

File: client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+   */
+  constructor(editor, buttons) {
+    this.mceIframe = document.getElementById(`${editor.id}_ifr`);
+    this.container = editor.getContainer();
+    this.mceToolbar = null;
+    this.mceStatusbar = null;
+
+    if (this.container) {
+      this.mceToolbar = tinymce.$('.mce-toolbar-grp', this.container)[0];
+      this.mceStatusbar = tinymce.$('.mce-statusbar', this.container)[0];
```
Comment (flamerohr) [inline client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js]: we could also use `this.container.querySelector('.className');` which selects the first instance element without coupling to jQuery.
I didn't know jQuery was a part of tinymce.

File: client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+  const toolbar = new TinymceInlineToolbar(editor, buttons);
+  let currentSelection = null;
+  let scrollTimeoutId = null;
+  let focused = false;
+
+  toolbar
+    .hide()
+    .renderTo(window.document.body);
+
+  toolbar.setStyles({ position: 'absolute' });
```
Comment (flamerohr) [inline client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js]: could this be part of the css file?
maybe not, it's a little bit difficult to track

File: client/src/legacy/TinyMCE_sslink.js [hunk 5]
```diff
@@ -23,11 +24,36 @@ const plugin = {
...
       text: 'Insert Link',
       menu: actions,
     });
+
+    function openLinkDialog() {
+      const node = tinymce.activeEditor.selection.getNode();
+      if (node.href) {
+        let command = 'sslinkexternal';
+
+        if (/^mailto:/.test(node.href)) {
```
Comment (phalkunz) [inline client/src/legacy/TinyMCE_sslink.js]: Good point! :) Will look into `TinyMCEActionRegistrar`

File: client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js (hunk 3 above)
Comment (phalkunz) [inline client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js]: 👍

File: client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js (hunk 4 above)
Comment (phalkunz) [inline client/src/components/TinymceInlineToolbar/TinymceInlineToolbar.js]: Actually it should be 👍

**Comment by dhensby** [comment]
this needs rebasing and re-building now

File: client/src/components/TinymceInlineToolbar/tests/TinymceInlineToolbar-test.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
+/* global jest, describe, beforeEach, it, expect, document */
+import { shouldShowToolbar, setupTinyMceInlineToolbar } from '../TinymceInlineToolbar';
+
+jest.unmock('react');
+jest.unmock('react-addons-test-utils');
```
Comment (flamerohr) [inline client/src/components/TinymceInlineToolbar/tests/TinymceInlineToolbar-test.js]: no longer need `jest.unmock` calls

File: client/src/components/TinymceInlineToolbar/tests/TinymceInlineToolbar-test.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            show: jest.genMockFunction(),
+            renderTo: jest.genMockFunction(),
+            getEl: () => document.createDocumentFragment(),
+          };
+        },
+      },
+    },
+  };
+
+  const editor = new Editor();
```
Comment (flamerohr) [inline client/src/components/TinymceInlineToolbar/tests/TinymceInlineToolbar-test.js]: Put `global.tinymce` and `editor` inside a `beforeEach(() => {` block, that way we can ignore mutations and possible side-effects by rebuilding this objects after each `it()` test.

We could mock `editor` with an object that contains the methods we expect to be used, e.g. `on = jest.fn()` and `hide = jest.fn(() => editor)`.

File: client/src/lib/TinyMCEActionRegistrar.js [hunk 8]
```diff
@@ -27,6 +32,42 @@ class TinyMCEActionRegistrar {
...
+  getEditorCommandFromUrl(url) {
+    let command = 'sslinkexternal';
+
+    if (/^mailto:/.test(url)) {
+      command = 'sslinkemail';
+    } else {
+      const commands = Object.keys(this.editorCommandsToUrlTestsMap);
+      const matchedCmd = commands
+        .filter(cmd => this.editorCommandsToUrlTestsMap[cmd].test(url))
+        .shift();
```
Comment (flamerohr) [inline client/src/lib/TinyMCEActionRegistrar.js]: can replace `.filter().shift()` with `.find()`

File: client/src/lib/TinyMCEActionRegistrar.js [hunk 9]
```diff
@@ -27,6 +32,42 @@ class TinyMCEActionRegistrar {
...
+   * Gets an editor command name based on the given url
+   *
+   * @param {String} url Inserted url
+   * @return {String} Editor command name
+   */
+  getEditorCommandFromUrl(url) {
+    let command = 'sslinkexternal';
+
+    if (/^mailto:/.test(url)) {
+      command = 'sslinkemail';
```
Comment (flamerohr) [inline client/src/lib/TinyMCEActionRegistrar.js]: it's a little weird to have this hardcoded, could this be using the  `editorCommands` API as well?

File: client/src/lib/TinyMCEActionRegistrar.js [hunk 10]
```diff
@@ -27,6 +32,42 @@ class TinyMCEActionRegistrar {
...
+  }
+
+  /**
+   * Gets an editor command name based on the given url
+   *
+   * @param {String} url Inserted url
+   * @return {String} Editor command name
+   */
+  getEditorCommandFromUrl(url) {
+    let command = 'sslinkexternal';
```
Comment (flamerohr) [inline client/src/lib/TinyMCEActionRegistrar.js]: could we have a `getDefaultCommand`? So this could be tweaked/replaced in future?

**Pull request merged**